neo/core/dataobject.py

```diff
@@ -0,0 +1,154 @@
+            # TODO: Does this make sense?
+            if len(value) == 0:
+                value = np.ndarray(own_length)
+                val_length = own_length
+            else:
+                # Note: len(o) also works for np.ndarray, it then uses the outmost dimension,
+                # which is exactly the desired behaviour here
+                val_length = len(value)
+
+            if not own_length == val_length:
+                raise ValueError("Incorrect length of array annotation: {} != {}".format(val_length, own_length))
+
+            for element in value:
+                if isinstance(element, (list, np.ndarray)):
+                    raise ValueError("Array annotations should only be 1-dimensional")
+
+                # Perform regular check for elements of array or list
+                _check_annotations(element)
+
+            # Create arrays from lists, because array annotations should be numpy arrays
+            if isinstance(value, list):
+                value = np.array(value)
+
+        return value
+
+    def array_annotate(self, **array_annotations):
+
+        """
+        Add annotations (non-standardized metadata) as arrays to a Neo data object.
+
+        Example:
+
+        >>> obj.array_annotate(key1=[value00, value01, value02], key2=[value10, value11, value12])
+        >>> obj.key2[1]
+        value11
+        """
+
+        array_annotations = self._check_array_annotations(array_annotations)
+        self.array_annotations.update(array_annotations)
+
+    def rescale(self, units):
+
+        # Use simpler functionality, if nothing will be changed
+        dim = pq.quantity.validate_dimensionality(units)
+        if self.dimensionality == dim:
+            return self.copy()
+
+        obj = self.duplicate_with_new_data(signal=self.view(pq.Quantity).rescale(dim), units=dim)
+        obj.array_annotations = self.array_annotations
+        obj.segment = self.segment
+        return obj
+
+    def array_annotations_at_index(self, index):  # TODO: Should they be sorted by key (current) or index?
+
+        """
+        Return dictionary of array annotations at a given index or list of indices
+        :param index: int, list, numpy array: The index (indices) from which the annotations are extracted
+        :return: dictionary of values or numpy arrays containing all array annotations for given index
+
+        Example:
+        >>> obj.array_annotate(key1=[value00, value01, value02], key2=[value10, value11, value12])
+        >>> obj.array_annotations_at_index(1)
+        {key1=value01, key2=value11}
+        """
+
+        index_annotations = {}
+
+        # Use what is given as an index to determine the corresponding annotations,
+        # if not possible, numpy raises an Error
+        for ann in self.array_annotations.keys():
+            index_annotations[ann] = self.array_annotations[ann][index]
+
+        return index_annotations
+
```

Comment:
I am wondering here if in case of type(index)==int what we must do.
In catual implementation we get a scalar in array_annotation, waybe we should have a array of size 1 because the analogsignal.shape is (N,1) when slicing on channel dim.
So having array_annotations as a scalar is not very consistent.
@apdavison @JuliaSprenger  :What do you think ?